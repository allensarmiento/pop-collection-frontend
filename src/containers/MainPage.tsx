import { Component } from 'react';
import Header from '../components/Header';
import SearchBox from '../components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundary from '../components/ErrorBoundary';
import PopCardList from '../components/PopCardList';
import { Pop } from '../state/types';

type MainPageProps = {
  className?: string;
  onSearchChange: Function;
  onRequestPops: Function;
  searchField: string;
  pops: Array<Pop>;
  isPending: boolean;
};

class MainPage extends Component<MainPageProps> {
  componentDidMount() {
    this.props.onRequestPops();
  }

  filterPops = (): Array<Pop> => {
    const { pops, searchField } = this.props;
    return pops.filter((pop) => {
      return pop.name.toLowerCase().includes(searchField.toLowerCase());
    });
  }

  render() {
    const { className, onSearchChange, isPending } = this.props;

    return (
      <div className={className}>
        <Header />
        <SearchBox searchChange={onSearchChange} />
        <Scroll>
          {isPending ? <h1>Loading</h1> : (
            <ErrorBoundary>
              <PopCardList pops={this.filterPops()} />
            </ErrorBoundary>
          )}
        </Scroll>
      </div>
    );
  }
}

export default MainPage;
