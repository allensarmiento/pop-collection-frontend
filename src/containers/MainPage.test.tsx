import { shallow, ShallowWrapper } from 'enzyme';
import MainPage from './MainPage';

let wrapper: ShallowWrapper | any;

beforeEach(() => {
  const mockProps = {
    onRequestPops: jest.fn(),
    onSearchChange: jest.fn(),
    pops: [],
    searchField: '',
    isPending: false,
  };

  wrapper = shallow(<MainPage {...mockProps} />);
});

it('renders without crashing', () => {
  expect(wrapper).toMatchSnapshot();
});

it('filters pops', () => {
  const mockProps = {
    onRequestPops: jest.fn(),
    onSearchChange: jest.fn(),
    pops: [],
    searchField: '',
    isPending: false,
  };

  wrapper = shallow(<MainPage {...mockProps} />);
  expect(wrapper.instance().filterPops()).toEqual([]);
});

it('filters pops correctly with matching searchField', () => {
  const filteredPops = [{
    name: 'Name 1',
    color: '#fff',
    category: 'Category 1',
    image: '',
  }];
  const mockProps = {
    onRequestPops: jest.fn(),
    onSearchChange: jest.fn(),
    pops: [{
      name: 'Name 1',
      color: '#fff',
      category: 'Category 1',
      image: '',
    }],
    searchField: 'Name 1',
    isPending: false,
  };

  wrapper = shallow(<MainPage {...mockProps} />);
  expect(wrapper.instance().filterPops()).toEqual(filteredPops);
});

it('filters pops correctly with no matching searchField', () => {
  const mockProps = {
    onRequestPops: jest.fn(),
    onSearchChange: jest.fn(),
    pops: [{
      name: 'Name 1',
      color: '#fff',
      category: 'Category 1',
      image: '',
    }],
    searchField: 'Name 99',
    isPending: false,
  };

  wrapper = shallow(<MainPage {...mockProps} />);
  expect(wrapper.instance().filterPops()).toEqual([]);
});
