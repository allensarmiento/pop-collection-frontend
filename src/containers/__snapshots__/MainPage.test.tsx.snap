// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <Memo(Header) />
  <SearchBox
    searchChange={[MockFunction]}
  />
  <Scroll>
    <ErrorBoundary>
      <CardList
        pops={Array []}
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
